alicloud/data_source_alicloud_yundun_dbaudit_instances.go

```diff
@@ -0,0 +1,205 @@
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"user_vswitch_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"private_domain": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"public_domain": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+						"instance_status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"license_code": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"public_network_access": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAlicloudDbauditInstancesRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	request := yundun_dbaudit.CreateDescribeInstancesRequest()
+	request.PageSize = requests.NewInteger(PageSizeSmall)
+	request.CurrentPage = requests.NewInteger(1)
+	var instances []yundun_dbaudit.Instance
+
+	// get name Regex
+	var nameRegex *regexp.Regexp
+	if v, ok := d.GetOk("name_regex"); ok {
+		if r, err := regexp.Compile(v.(string)); err == nil {
+			nameRegex = r
+		}
+	}
+
+	if v, ok := d.GetOk("ids"); ok {
+		ids, _ := v.([]interface{})
+		var ids_str []string
+		for _, v_instance_id := range ids {
+			ids_str = append(ids_str, v_instance_id.(string))
+		}
+		request.InstanceId = &ids_str
+	}
+	for {
+		raw, err := client.WithDbauditClient(func(dbauditClient *yundun_dbaudit.Client) (interface{}, error) {
+			return dbauditClient.DescribeInstances(request)
+		})
+		if err != nil {
+			return WrapErrorf(err, DataDefaultErrorMsg, "alicloud_yundun_dbaudit", request.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		addDebug(request.GetActionName(), raw, request.RpcRequest, request)
+		response, _ := raw.(*yundun_dbaudit.DescribeInstancesResponse)
+		if len(response.Instances) < 1 {
+			break
+		}
+
+		for _, e := range response.Instances {
+			if nameRegex != nil && !nameRegex.MatchString(e.Description) {
+				continue
+			}
+			instances = append(instances, e)
+		}
+
+		if len(response.Instances) < PageSizeSmall {
+			break
+		}
+
+		currentPageNo := request.CurrentPage
+
+		if page, err := getNextpageNumber(currentPageNo); err != nil {
+			return WrapError(err)
+		} else {
+			request.CurrentPage = page
+		}
+	}
+
+	var instanceIds []string
+	for _, instance := range instances {
+		instanceIds = append(instanceIds, instance.InstanceId)
+	}
+	if len(instanceIds) < 1 {
+		return WrapError(extractDbauditInstance(d, nil))
+	}
+	var specs []yundun_dbaudit.InstanceAttribute
+	for _, instanceId := range instanceIds {
+		request := yundun_dbaudit.CreateDescribeInstanceAttributeRequest()
+		request.InstanceId = instanceId
+		raw, err := client.WithDbauditClient(func(dbauditClient *yundun_dbaudit.Client) (interface{}, error) {
+			return dbauditClient.DescribeInstanceAttribute(request)
```

Comment:
Because the result of DescribeInstance is a subset of DescribeInstanceAttribute，using DescribeInstanceAttribute offers more flexibility for further extension(more attributes required).

Reply:
will do